akka-projection-jdbc/src/main/scala/akka/projection/jdbc/internal/JdbcOffsetStore.scala

```diff
@@ -0,0 +1,115 @@
+
+  def createIfNotExists(): Future[Done] = {
+    withConnection(jdbcSessionFactory) { conn =>
+      tryWithResource(conn.createStatement()) { stmt =>
+        stmt.execute(settings.dialect.createTableStatement)
+        stmt.execute(settings.dialect.alterTableStatement)
+        Done
+      }
+    }
+  }
+
+  def clearOffset(projectionId: ProjectionId): Future[Done] =
+    withConnection(jdbcSessionFactory) { conn =>
+      tryWithResource(conn.prepareStatement(settings.dialect.clearOffsetStatement)) { stmt =>
+        stmt.setString(1, projectionId.name)
+        stmt.setString(2, projectionId.key)
+        stmt.executeUpdate()
+        Done
+      }
+    }
+
+  def readOffset[Offset](projectionId: ProjectionId): Future[Option[Offset]] =
+    withConnection(jdbcSessionFactory) { conn =>
+      // init Statement in try-with-resource
+      tryWithResource(conn.prepareStatement(settings.dialect.readOffsetQuery)) { stmt =>
+
+        stmt.setString(1, projectionId.name)
+        stmt.setString(2, projectionId.key)
+
+        // init ResultSet in try-with-resource
+        tryWithResource(stmt.executeQuery()) { resultSet =>
+          // FIXME: need to iterate over all results to build MultipleOffsets
+          if (resultSet.first()) {
+            val offsetStr = resultSet.getString("OFFSET")
+            val manifest = resultSet.getString("MANIFEST")
+            Some(fromStorageRepresentation(offsetStr, manifest))
+          } else None
+        }
+      }
+    }
+
+  /**
+   * Like saveOffset, but async. Useful for resetting an offset
+   */
+  def saveOffset[Offset](projectionId: ProjectionId, offset: Offset): Future[Done] =
+    withConnection(jdbcSessionFactory) { conn =>
+      saveOffsetBlocking(conn, projectionId, offset)
+    }
+
+  /**
+   * This method is explicitly made non-async because it's used together
+   * with the users' handler code and run inside the same Future
+   */
+  def saveOffsetBlocking[Offset](conn: Connection, projectionId: ProjectionId, offset: Offset): Done = {
+
+    tryWithResource(conn.prepareStatement(settings.dialect.insertOrUpdateStatement)) { stmt =>
+
+      toStorageRepresentation(projectionId, offset) match {
+        case SingleOffset(id, manifest, offsetStr, mergeable) =>
+          stmt.setString(InsertIndices.PROJECTION_NAME, id.name)
+          stmt.setString(InsertIndices.PROJECTION_KEY, id.key)
+          stmt.setString(InsertIndices.OFFSET, offsetStr)
+          stmt.setString(InsertIndices.MANIFEST, manifest)
+          stmt.setBoolean(InsertIndices.MERGEABLE, mergeable)
+          val now: Instant = Instant.now(clock)
+          stmt.setTimestamp(InsertIndices.LAST_UPDATED, new Timestamp(now.toEpochMilli))
+
+        case _ =>
+          // FIXME: add support for MultipleOffsets
+          throw new IllegalArgumentException("The JdbcOffsetStore does not currently support MergeableOffset")
+      }
+
+      stmt.executeUpdate()
```

Comment:
In another PR I think we should give it a try if we can get away with a plain update statement and if that returns 0 rows we execute a plain insert statement. I think that is much more standardized across database dialects.

Reply:
Yes, once we finish with the delivery strategies impl, we can re-work the dialect part. Ideally, we should even not have the concept of a dialect.